File: backend/morphologie-api/src/main/java/com/morphologie/api/service/SchemeService.java
package com.morphologie.api.service;

import com.morphologie.engine.Morphologie;
import com.morphologie.api.model.SchemeModel;
import com.morphologie.api.repository.SchemeRepository;
import com.morphologie.api.exception.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;

// ============= Scheme Service =============
@Service
public class SchemeService {
    
    private final SchemeRepository schemeRepository;
    private final Morphologie engine;
    
    public SchemeService(SchemeRepository schemeRepository, Morphologie engine) {
        this.schemeRepository = schemeRepository;
        this.engine = engine;
    }
    
    /**
     * Add a new scheme
     */
    @Transactional
    public SchemeModel addScheme(String nom, String type, String description) {
        // Validate
        if (nom == null || nom.trim().isEmpty()) {
            throw new InvalidSchemeException("Scheme name cannot be empty");
        }
        
        // Validate contains ف ع ل
        if (!nom.contains("ف") || !nom.contains("ع") || !nom.contains("ل")) {
            throw new InvalidSchemeException("Scheme must contain ف ع ل");
        }
        
        // Check if already exists
        if (schemeRepository.existsByNom(nom)) {
            throw new SchemeAlreadyExistsException(nom);
        }
        
        // Add to engine
        engine.ajouterScheme(nom);
        
        // Save to MongoDB
        SchemeModel scheme = new SchemeModel(nom, nom, true, type != null ? type : "CUSTOM");
        scheme.setDescription(description);
        return schemeRepository.save(scheme);
    }
    
    /**
     * Get all schemes
     */
    public List<SchemeModel> getAllSchemes() {
        return schemeRepository.findAll();
    }
    
    /**
     * Get scheme by name
     */
    public SchemeModel getScheme(String nom) {
        return schemeRepository.findByNom(nom)
                .orElseThrow(() -> new SchemeNotFoundException(nom));
    }
    
    /**
     * Get schemes by type
     */
    public List<SchemeModel> getSchemesByType(String type) {
        return schemeRepository.findByType(type);
    }
    
    /**
     * Modify an existing scheme
     */
    @Transactional
    public SchemeModel modifyScheme(String nom, String newPattern) {
        SchemeModel scheme = getScheme(nom);
        
        // Validate new pattern
        if (!newPattern.contains("ف") || !newPattern.contains("ع") || !newPattern.contains("ل")) {
            throw new InvalidSchemeException("New pattern must contain ف ع ل");
        }
        
        // Modify in engine
        engine.modifierScheme(nom, newPattern);
        
        // Update in MongoDB
        scheme.setNom(newPattern);
        scheme.setPattern(newPattern);
        return schemeRepository.save(scheme);
    }
    
    /**
     * Delete a scheme
     */
    @Transactional
    public void deleteScheme(String nom) {
        if (!schemeRepository.existsByNom(nom)) {
            throw new SchemeNotFoundException(nom);
        }
        
        // Remove from engine
        engine.supprimerScheme(nom);
        
        // Delete from MongoDB
        schemeRepository.deleteByNom(nom);
    }
    
    /**
     * Get most popular schemes
     */
    public List<SchemeModel> getPopularSchemes() {
        return schemeRepository.findTop10ByOrderByUsageCountDesc();
    }
    
    /**
     * Get total count of schemes
     */
    public long getSchemeCount() {
        return schemeRepository.count();
    }
}

File: Frontend/services/schemeService.js
// services/schemeService.js - نسخة محسنة ✅
import api from './api';

export const schemeService = {
  // جلب كل الأوزان
  getAllSchemes: async () => {
    try {
      console.log('📤 Fetching all schemes...');
      const response = await api.get('/schemes');
      console.log('📥 getAllSchemes response:', response);
      
      // التأكد من أن response عبارة عن مصفوفة
      if (Array.isArray(response)) {
        return response;
      } else if (response && response.data && Array.isArray(response.data)) {
        return response.data;
      } else {
        console.warn('⚠️ Unexpected response format:', response);
        return [];
      }
    } catch (error) {
      console.error('❌ getAllSchemes error:', error);
      throw error;
    }
  },
  
  // جلب وزن معين
  getScheme: async (nom) => {
    try {
      console.log('📤 Fetching scheme:', nom);
      const response = await api.get(`/schemes/${encodeURIComponent(nom)}`);
      return response;
    } catch (error) {
      console.error('❌ getScheme error:', error);
      throw error;
    }
  },
  
  // جلب الأوزان حسب النوع
  getSchemesByType: async (type) => {
    try {
      console.log('📤 Fetching schemes by type:', type);
      const response = await api.get(`/schemes/types/${type}`);
      return response;
    } catch (error) {
      console.error('❌ getSchemesByType error:', error);
      throw error;
    }
  },
  
  // إضافة وزن جديد
  addScheme: async (schemeData) => {
    try {
      console.log('📤 Adding scheme:', schemeData);
      
      // التأكد من صحة البيانات
      if (!schemeData.nom || !schemeData.nom.trim()) {
        throw new Error('اسم النمط مطلوب');
      }
      
      const response = await api.post('/schemes', {
        nom: schemeData.nom.trim(),
        type: schemeData.type || 'CUSTOM',
        description: schemeData.description || ''
      });
      
      console.log('📥 Add scheme response:', response);
      return response;
    } catch (error) {
      console.error('❌ addScheme error:', error.response || error);
      throw error;
    }
  },
  
  // تعديل وزن
  modifyScheme: async (oldNom, newPattern) => {
    try {
      console.log('📤 Modifying scheme:', oldNom, '→', newPattern);
      
      if (!oldNom || !newPattern) {
        throw new Error('الاسم القديم والجديد مطلوبان');
      }
      
      const response = await api.put(`/schemes/${encodeURIComponent(oldNom)}`, { 
        newPattern: newPattern.trim() 
      });
      
      console.log('📥 Modify scheme response:', response);
      return response;
    } catch (error) {
      console.error('❌ modifyScheme error:', error.response || error);
      throw error;
    }
  },
  
  // حذف وزن
  deleteScheme: async (nom) => {
    try {
      console.log('🗑️ Deleting scheme:', nom);
      
      if (!nom || !nom.trim()) {
        throw new Error('اسم النمط مطلوب للحذف');
      }
      
      const response = await api.delete(`/schemes/${encodeURIComponent(nom)}`);
      console.log('📥 Delete scheme response:', response);
      return response;
    } catch (error) {
      console.error('❌ deleteScheme error:', error.response || error);
      throw error;
    }
  },
  
  // الأوزان الأكثر استخداماً
  getPopularSchemes: async (limit = 10) => {
    try {
      console.log('📤 Fetching popular schemes, limit:', limit);
      const response = await api.get('/schemes/stats/popular', {
        params: { limit }
      });
      return response;
    } catch (error) {
      console.error('❌ getPopularSchemes error:', error);
      throw error;
    }
  },
  
  // البحث عن الأوزان
  searchSchemes: async (query) => {
    try {
      console.log('📤 Searching schemes:', query);
      
      if (!query || !query.trim()) {
        return [];
      }
      
      const response = await api.get('/schemes/search', {
        params: { q: query.trim() }
      });
      
      return response;
    } catch (error) {
      console.error('❌ searchSchemes error:', error);
      throw error;
    }
  },
  
  // إحصائيات الأوزان
  getSchemeStats: async () => {
    try {
      console.log('📤 Fetching scheme stats...');
      const response = await api.get('/schemes/stats');
      return response;
    } catch (error) {
      console.error('❌ getSchemeStats error:', error);
      throw error;
    }
  },
  
  // التحقق من وجود وزن
  checkSchemeExists: async (nom) => {
    try {
      console.log('🔍 Checking if scheme exists:', nom);
      const response = await api.get(`/schemes/${encodeURIComponent(nom)}/exists`);
      return response;
    } catch (error) {
      console.error('❌ checkSchemeExists error:', error);
      throw error;
    }
  }
};